### ISSUE TITLE
Request support for RWKV-4-World model.

### ISSUE THREAD

**Issue by SetoKaiba** [huggingface/transformers#24842]
### Model description

As RWKV-4-World is using a different tokenizer and vocabs, the current RWKV support in transformers is incompatible.

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

https://huggingface.co/StarRing2022/RWKV-4-World-1.5B
@StarRing2022

**Comment by SetoKaiba** [comment]
@sgugger Is it possible to add the support for it? I'd like to use PEFT to fine tune the model.

**Comment by StarRing2022** [comment]
We have implemented the function of fine-tuning the RWKV-World model using the peft library, and we will immediately publish it in HF

**Comment by StarRing2022** [comment]
The relevant code will be provided under this issue

**Comment by StarRing2022** [comment]
Please refer to：
https://github.com/StarRing2022/HF-For-RWKVWorld-LoraAlpaca
https://huggingface.co/StarRing2022/RWKV-4-World-7B

**Comment by SetoKaiba** [comment]
Thank you for your great work.
Will you make a PR for it? Or will you just keep it a separate repo?
I'd like to see it PR back to hf transformers. And others can get your work easily.

https://github.com/StarRing2022/RingRWKV
I think this should be mentioned in the example github repo as well.

**Comment by SetoKaiba** [comment]
> Please refer to： https://github.com/StarRing2022/HF-For-RWKVWorld-LoraAlpaca https://huggingface.co/StarRing2022/RWKV-4-World-7B

@StarRing2022 Thank you for your great work.
Will you make a PR for it? Or will you just keep it a separate repo?
I'd like to see it PR back to hf transformers. And others can get your work easily.

https://github.com/StarRing2022/RingRWKV
I think this should be mentioned in the example github repo as well.

**Comment by sgugger** [comment]
cc @younesbelkada

**Comment by StarRing2022** [comment]
Thanks，there are indeed some issues with the RWKV in HF format on the official transformers. Recently, we have found that there are issues such as CFG and sample_ Logits, PR is necessary, and we also hope that the official can provide improvements

**Comment by john1089** [comment]
I think that hf transformers is working with RWKV world ,just use the ringrwkv tokenizer(or the world tokenizer in the official repo) and it will work

**Comment by StarRing2022** [comment]
Yes,We are trying to contact with HF transformers

**Comment by qgallouedec** [comment]
Hey @SetoKaiba, sorry for the very slow reply here, this fell through the cracks while we were catching up on the backlog.

The root blocker isn't the model architecture RWKV-4 is already supported, see https://github.com/huggingface/transformers/blob/96fe6dce36cc929a5ffd3e34296554c4cb6b669e/src/transformers/models/rwkv/modeling_rwkv.py but the World tokenizer's vocab, which ships as a `.txt` file rather than the JSON format `PreTrainedTokenizerFast` expects.
The conversion script already accepts a prebuilt tokenizer file https://github.com/huggingface/transformers/blob/96fe6dce36cc929a5ffd3e34296554c4cb6b669e/src/transformers/models/rwkv/convert_rwkv_checkpoint_to_hf.py#L80-L90 and folks in the duplicate thread (#25253) shared a script that builds exactly that JSON vocab/merges pair from BlinkDL's rwkv_vocab txt file.

Given this has had zero +1s in three years and the workaround is known, I'm closing as won't implement for now rather than carrying it as a native port, thanks for the request though!
